gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Client.java

```diff
@@ -521,25 +523,31 @@ protected void initializeImplementation() {
             // the cost of this single threaded executor here should be fairly small because it is only used once at
             // initialization and shutdown. since users will typically construct a Client once for the life of their
             // application there shouldn't be tons of thread pools being created and destroyed.
-            final BasicThreadFactory threadFactory = new BasicThreadFactory.Builder().namingPattern("gremlin-driver-initializer").build();
+            final BasicThreadFactory threadFactory =
+                    new BasicThreadFactory.Builder().namingPattern("gremlin-driver-initializer").build();
             final ExecutorService hostExecutor = Executors.newSingleThreadExecutor(threadFactory);
 
-            try {
-                CompletableFuture.allOf(cluster.allHosts().stream()
-                        .map(host -> CompletableFuture.runAsync(() -> initializeConnectionSetupForHost.accept(host), hostExecutor))
-                        .toArray(CompletableFuture[]::new))
-                        .join();
-            } catch (CompletionException ex) {
-                Throwable cause;
-                Throwable result = ex;
-                if (null != (cause = ex.getCause())) {
-                    result = cause;
-                }
+            // hosts are not marked as available at Cluster initialization, so we are marking hosts as available here
+            // if connection pools can be initialized successfully.
+            final List<Pair<Host, CompletableFuture<Void>>> completableFutures = new ArrayList<>();
+            for (Host host : cluster.allHosts()) {
+                completableFutures.add(new ImmutablePair<>(host,
+                        CompletableFuture.runAsync(() -> initializeConnectionSetupForHost.accept(host), hostExecutor)));
+            }
 
-                logger.error("", result);
-            } finally {
-                hostExecutor.shutdown();
+            for (Pair<Host, CompletableFuture<Void>> completableFuture : completableFutures) {
+                try {
+                    completableFuture.getRight().join();
+                    completableFuture.getLeft().makeAvailable();
+                } catch (CompletionException ex) {
+                    Throwable result = ex;
```

Comment:
This block can be simplified to one line.

`logger.error("", (ex.getCause() == null) ? ex : ex.getCause());`

Reply:
Okay, I'll change it.